storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -4017,122 +4014,1889 @@ dberr_t fil_space_t::garbage_collect(bool shutdown)
+    dberr_t err= DB_SUCCESS;
+    if (left_page_no != FIL_NULL)
+    {
+      m_left_block= fsp_get_latched_page(page_id_t{0, left_page_no},
+                                         m_mtr, &err);
+      ut_ad(!m_left_block == (err != DB_SUCCESS));
+      if (!m_left_block)
+        return err;
+    }
+
+    uint32_t right_page_no= btr_page_get_next(m_cur_block->page.frame);
+    if (right_page_no != FIL_NULL)
+    {
+      m_right_block= fsp_get_latched_page(page_id_t{0, right_page_no},
+                                          m_mtr, &err);
+      ut_ad(!m_right_block == (err != DB_SUCCESS));
+      if (!m_right_block)
+        return err;
+    }
+
+    m_parent_block= fsp_get_latched_page(page_id_t{0, parent_page},
+                                         m_mtr, &err);
+    return err;
   }
-  uint32_t fixed_size= srv_sys_space.get_min_size(),
-           header_size= space->size_in_header;
-  mtr.commit();
 
-  if (last_used_extent >= header_size || fixed_size >= header_size)
-    /* Tablespace is being used within fixed size */
-    return;
+  /** Modify the FIL_PAGE_NEXT, FIL_PAGE_PREV, CHILD_PAGE of
+  respective left, right and parent block to new page number */
+  void complete(uint32_t new_page_no, uint32_t parent_offset) noexcept
+  {
+    if (m_left_block)
+      m_mtr->write<4>(*m_left_block,
+                      m_left_block->page.frame + FIL_PAGE_NEXT,
+                      new_page_no);
+
+    if (m_right_block)
+      m_mtr->write<4>(*m_right_block,
+                      m_right_block->page.frame + FIL_PAGE_PREV,
+                      new_page_no);
+
+    m_mtr->write<4>(*m_parent_block,
+                    m_parent_block->page.frame + parent_offset,
+                    new_page_no);
+  }
+};
 
-  /* Set fixed size as threshold to truncate */
-  if (fixed_size > last_used_extent)
-    last_used_extent= fixed_size;
+/** page operation for the system tablespace does the 2 things:
+1) Page Allocation
+2) Page removal
 
-  bool old_dblwr_buf= buf_dblwr.in_use();
-  /* Flush all pages in buffer pool, so that it doesn't have to
-  use doublewrite buffer and disable dblwr and there should
-  be enough space in redo log */
-  log_make_checkpoint();
-  fil_system.set_use_doublewrite(false);
+Steps for page allocation depends on new extent state.
 
-  buf_block_t *header= nullptr;
-  ut_ad(!fsp_tablespace_validate(space));
+(1) If the xdes_get_state(new_descr) == XDES_FREE then
+remove the new extent from FSP_FREE list
 
-  mtr.start();
-  mtr.x_lock_space(space);
+    (1.1) If the page has to be allocated for segment then
+          add the newly allocated extent descriptor to
+          FSEG_NOT_FULL list and make the xdes_set_state(new_descr)
+          as XDES_FSEG
+
+    (1.2) If the page has to be non-segment page then add the
+          newly allocated extent descriptor to FSP_FREE_FRAG list
+          and make the xdes_set_state(new_descr) as XDES_FREE_FRAG
+
+    (1.3) return the free page from the new extent
```

Comment:
"return" is a bit too overloaded term here. I believe that you mean "allocate a page from the new extent". Initially I was actually misinterpreting it as "free the page to the new extent", but that would really make no sense, because this is describing the steps to allocate a page.